src/transformers/models/nucleus_x/modeling_nucleus_x.py

```diff
@@ -0,0 +1,1449 @@
+    """Drop paths (Stochastic Depth) per sample (when applied in main path of residual blocks).
+
+    This is the same as the DropConnect impl I created for EfficientNet, etc networks, however, the original name is
+    misleading as 'Drop Connect' is a different form of dropout in a separate paper... See discussion:
+    https://github.com/tensorflow/tpu/issues/494#issuecomment-532968956 ... I've opted for changing the layer and
+    argument names to 'drop path' rather than mix DropConnect as a layer name and use 'survival rate' as the argument.
+
+    """
+    if drop_prob == 0.0 or not training:
+        return x
+    keep_prob = 1 - drop_prob
+    shape = (x.shape[0],) + (1,) * (x.ndim - 1)  # work with diff dim tensors, not just 2D ConvNets
+    random_tensor = x.new_empty(shape).bernoulli_(keep_prob)
+    if keep_prob > 0.0 and scale_by_keep:
+        random_tensor.div_(keep_prob)
+    return x * random_tensor
+
+
+class NucleusXDropPath(nn.Module):
+    """Drop paths (Stochastic Depth) per sample (when applied in main path of residual blocks).
+    Copied from https://github.com/microsoft/torchscale/blob/main/torchscale/component/droppath.py"""
+
+    def __init__(self, drop_prob=None):
+        super().__init__()
+        self.drop_prob = drop_prob
+
+    def forward(self, x):
+        return drop_path(x, self.drop_prob, self.training)
+
+    def extra_repr(self):
+        return "p={}".format(self.drop_prob)
+
+
+class NucleusXDecoderLayer(nn.Module):
+    """Decoder layer for NucleusX.
+    Adapted from https://github.com/microsoft/torchscale/blob/main/torchscale/architecture/retnet.py
+    """
+
+    def __init__(self, config: NucleusXConfig, depth: int):
+        super().__init__()
+        self.config = config
+        self.embed_dim = config.decoder_embed_dim
+        self.dropout_module = nn.Dropout(config.dropout)
+
+        if config.drop_path_rate > 0:
+            drop_path_prob = np.linspace(0, config.drop_path_rate, config.decoder_layers)[depth]
+            self.drop_path = NucleusXDropPath(drop_path_prob)
+        else:
+            self.drop_path = None
+
+        self.retention = NucleusXMultiScaleRetention(config)
+
+        self.normalize_before = config.decoder_normalize_before
+
+        self.retention_layer_norm = build_rms_norm(self.embed_dim, eps=config.rms_norm_eps)
+
+        self.ffn_dim = config.decoder_ffn_embed_dim
+
+        self.ffn = self.build_ffn()
+
+        self.final_layer_norm = build_rms_norm(self.embed_dim, eps=config.rms_norm_eps)
+
+        if config.deepnorm:
+            self.alpha = math.pow(2.0 * config.decoder_layers, 0.25)
+        else:
+            self.alpha = 1.0
+
+    def build_ffn(self):
+        return NucleusXGLU(
+            self.embed_dim,
+            self.ffn_dim,
+            self.config.activation_fn,
+            self.config.dropout,
+            self.config.activation_dropout,
+        )
```

Comment:
I have a very slight preference to remove this method and just inline it in the init: `self.ffn = NucleusXGLU(...)`